File: py/graph and trees/particular_area_in_city_bfs_dfs.py
from collections import deque

# Define nodes and mapping for matrix
nodes = ['A','B','C','D','E']
index = {nodes[i]: i for i in range(len(nodes))}
n = len(nodes)

# --- Adjacency List for BFS ---
adj_list = {
    'A': ['B','C'],
    'B': ['A','D'],
    'C': ['A','D','E'],
    'D': ['B','C','E'],
    'E': ['C','D']
}

def bfs(start):
    visited = {node: False for node in nodes}
    order = []
    q = deque([start])
    visited[start] = True

    while q:
        u = q.popleft()
        order.append(u)
        for v in adj_list[u]:
            if not visited[v]:
                visited[v] = True
                q.append(v)
    return order


# --- Adjacency Matrix for DFS ---
adj_matrix = [[0]*n for _ in range(n)]
edges = [('A','B'),('A','C'),('B','D'),('C','D'),('C','E'),('D','E')]

for u,v in edges:
    i,j = index[u], index[v]
    adj_matrix[i][j] = adj_matrix[j][i] = 1  # undirected

visited_matrix = [False]*n
dfs_order = []

def dfs_matrix(u_idx):
    visited_matrix[u_idx] = True
    dfs_order.append(nodes[u_idx])
    for v_idx in range(n):
        if adj_matrix[u_idx][v_idx] == 1 and not visited_matrix[v_idx]:
            dfs_matrix(v_idx)


# --- Run BFS and DFS ---
start = 'A'
print("BFS order:", ' '.join(bfs(start)))

dfs_order.clear()
visited_matrix = [False]*n
dfs_matrix(index[start])
print("DFS order:", ' '.join(dfs_order))

File: cpp/graph and tree/particular_area_in_city_bfs_dfs.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <functional>
using namespace std;

// Helper: map A,B,C... to 0,1,2,...
int idx(char c){ return c - 'A'; }

int main() {
    int n = 5; // nodes: A..E
    vector<vector<int>> adjList(n);

    // edges (undirected)
    vector<pair<char,char>> edges = {
        {'A','B'},{'A','C'},{'B','D'},{'C','D'},{'C','E'},{'D','E'}
    };

    // Build adjacency list
    for (auto &e: edges) {
        int u = idx(e.first), v = idx(e.second);
        adjList[u].push_back(v);
        adjList[v].push_back(u);
    }

    // --- BFS using adjacency list ---
    vector<bool> vis(n,false);
    queue<int> q;
    vector<char> bfs_order;
    int start = idx('A');
    vis[start] = true;
    q.push(start);

    while (!q.empty()) {
        int u = q.front(); q.pop();
        bfs_order.push_back('A' + u);
        for (int v : adjList[u]) {
            if (!vis[v]) {
                vis[v] = true;
                q.push(v);
            }
        }
    }

    cout << "BFS order: ";
    for (char c : bfs_order) cout << c << ' ';
    cout << endl;


    // --- Build adjacency matrix for DFS ---
    vector<vector<int>> adjMat(n, vector<int>(n,0));
    for (auto &e: edges) {
        int u = idx(e.first), v = idx(e.second);
        adjMat[u][v] = adjMat[v][u] = 1;
    }

    vector<bool> visited(n,false);
    vector<char> dfs_order;

    function<void(int)> dfs = [&](int u){
        visited[u] = true;
        dfs_order.push_back('A' + u);
        for (int v = 0; v < n; ++v) {
            if (adjMat[u][v] && !visited[v])
                dfs(v);
        }
    };

    dfs(start);

    cout << "DFS order: ";
    for (char c : dfs_order) cout << c << ' ';
    cout << endl;

    return 0;
}
